src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs

```diff
@@ -0,0 +1,228 @@
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Localization;
+using Microsoft.Extensions.Logging;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Records;
+using OrchardCore.DisplayManagement;
+using OrchardCore.Entities;
+using OrchardCore.Navigation;
+using OrchardCore.Search.Abstractions;
+using OrchardCore.Search.Abstractions.ViewModels;
+using OrchardCore.Search.Model;
+using OrchardCore.Security.Permissions;
+using OrchardCore.Settings;
+using YesSql;
+using YesSql.Services;
+
+namespace OrchardCore.Search;
+
+public class SearchController : Controller
+{
+    private readonly IAuthorizationService _authorizationService;
+    private readonly ISiteService _siteService;
+    private readonly ISession _session;
+    private readonly IStringLocalizer S;
+    private readonly IServiceProvider _serviceProvider;
+    private readonly IShapeFactory _shapeFactory;
+    private readonly dynamic New;
+    private readonly ILogger _logger;
+
+    public SearchController(
+        IAuthorizationService authorizationService,
+        ISiteService siteService,
+        ISession session,
+        IStringLocalizer<SearchController> stringLocalizer,
+        IEnumerable<IPermissionProvider> permissionProviders,
+        IServiceProvider serviceProvider,
+        IShapeFactory shapeFactory,
+        ILogger<SearchController> logger
+        )
+    {
+        _authorizationService = authorizationService;
+        _siteService = siteService;
+        _session = session;
+        S = stringLocalizer;
+        _serviceProvider = serviceProvider;
+        _shapeFactory = shapeFactory;
+        New = shapeFactory;
+        _logger = logger;
+    }
+
+    public async Task<IActionResult> Search(string index, SearchIndexViewModel viewModel, PagerSlimParameters pagerParameters)
+    {
+        if (!String.IsNullOrEmpty(index))
+        {
+            viewModel.Index = index;
+        }
+
+        var searchServices = _serviceProvider.GetServices<ISearchService>();
+
+        if (!searchServices.Any())
+        {
+            _logger.LogInformation("No search provider feature is enabled.");
+
+            return BadRequest("Missing search provider service.");
+        }
+
+        var site = await _siteService.GetSiteSettingsAsync();
+
+        var searchSettings = site.As<SearchSettings>();
+
+        var searchService = GetSearchService(searchServices, searchSettings?.SearchProviderAreaName);
+
+        var indexName = !String.IsNullOrWhiteSpace(viewModel.Index) ? viewModel.Index.Trim() : await searchService.DefaultIndexAsync();
+
+        if (indexName == null || !await searchService.ExistsAsync(indexName))
+        {
+            _logger.LogInformation("Couldn't execute search. The search index doesn't exist.");
+
+            return BadRequest("Search provider is not configured.");
+        }
+
+        var permission = await searchService.GetPermissionAsync(indexName);
```

Comment:
Search permission, common to all providers -> abstraction

Reply:
Done!